### PR TITLE
Improve our use of unsafe code

### PR DESCRIPTION
**Pull request by nrc**
###  What have you changed?

I've reduced our unsafe footprint by removing or restricting some uses of `unsafe`. I've documented some unsafe usage. I've changed some uses of `unsafe` in minor ways to better reflect the actual unsafeness and make it easier to check the necessary invariants. This is not a full audit of our unsafe code, just a small step.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

N/A

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

PTAL @breeswish @5kbpers

### CODE REVIEW

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 1]
```diff
@@ -98,12 +98,14 @@ fn ymd_hms_nanos<T: TimeZone>(
         })
 }
 
+// Safety: caller must ensure `bs` is valid utf8.
```
Comment (hicqu) [inline components/tidb_query/src/codec/mysql/time/mod.rs]: The comment conflicts with the new implementation.

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 2]
```diff
@@ -98,12 +98,14 @@ fn ymd_hms_nanos<T: TimeZone>(
...
 #[inline]
-fn from_bytes(bs: &[u8]) -> &str {
-    unsafe { str::from_utf8_unchecked(bs) }
+unsafe fn from_bytes(bs: &[u8]) -> &str {
+    str::from_utf8_unchecked(bs)
 }
 
-fn split_ymd_hms_with_frac_as_s(
+// Safety: caller must ensure `s` and `frac` are valid ascii.
```
Comment (hicqu) [inline components/tidb_query/src/codec/mysql/time/mod.rs]: The comment is also stale.

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 1 above)
Comment (nrc) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: Where is the new implementation? I don't see a conflict here

**Comment by siddontang** [comment]
PTAL @breeswish

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 1 above)
Comment (nrc) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: ping @hicqu

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 2 above)
Comment (nrc) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: I've updated the wording on this one.

File: components/tidb_query/src/batch/executors/util/mod.rs [hunk 3]
```diff
@@ -40,9 +39,13 @@ pub unsafe fn eval_exprs_decoded_no_lifetime<'a>(
     input_logical_rows: &[usize],
     output: &mut Vec<RpnStackNode<'a>>,
 ) -> Result<()> {
+    unsafe fn erase_lifetime<'a, T: ?Sized>(v: &T) -> &'a T {
```
Comment (5kbpers) [inline components/tidb_query/src/batch/executors/util/mod.rs]: Is it better to add `#[inline]` here?

Comment (nrc) [reply components/tidb_query/src/batch/executors/util/mod.rs]: I'm 99% sure that rustc will inline this

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 1 above)
Comment (hicqu) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: I mean if we mark the function as `unsafe`, the comment is not necessary any more, I guess.

Comment (nrc) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: Ah, I see. I think that if the comment is safe, then the caller can assume that calling the function is safe for any inputs. If the function is unsafe, that means that the caller must ensure some invariant, the comment states what that invariant is.

File: src/storage/kv/mod.rs [hunk 4]
```diff
@@ -710,27 +710,44 @@ thread_local! {
...
+        if !(*e.get()).is_null() {
+            let engine = Box::into_raw(Box::new(engine)) as *mut ();
+            *e.get() = engine;
+        }
+    });
 }
 
 /// Destroy the thread local engine.
+///
+/// Postcondition: `TLS_ENGINE_ANY` is null.
```
Comment (breeswish) [inline src/storage/kv/mod.rs]: I believe this function is unsafe, for example, incorrect destructor will be called (which will lead to memory corruption) if the supplied `E` is not the same as the one in `set_tls_engine`.

Comment (nrc) [reply src/storage/kv/mod.rs]: Good catch!

Comment (nrc) [reply src/storage/kv/mod.rs]: addressed

**Comment by nrc** [comment]
ping @hicqu are your comments addressed?

**Pull request merged**